contracts/cosmwasm-vm/cw-centralized-connection/src/contract.rs

```diff
@@ -0,0 +1,213 @@
+use cw_xcall_lib::network_address::NetId;
+
+use super::*;
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cw-mock-dapp";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+impl<'a> CwCentralizedConnection<'a> {
+    pub fn instantiate(
+        &mut self,
+        deps: DepsMut,
+        _env: Env,
+        _info: MessageInfo,
+        msg: InstantiateMsg,
+    ) -> Result<Response, ContractError> {
+        set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+        let relayer = deps.api.addr_validate(&msg.relayer)?;
+        self.store_admin(deps.storage, relayer)?;
+
+        let xcall_address = deps.api.addr_validate(&msg.xcall_address)?;
+        self.store_xcall(deps.storage, xcall_address)?;
+        self.store_denom(deps.storage, msg.denom)?;
+
+        let _ = self.store_conn_sn(deps.storage, 0);
+
+        Ok(Response::new()
+            .add_attribute("action", "instantiate")
+            .add_attribute("relayer", msg.relayer)
+            .add_attribute("xcall_address", msg.xcall_address))
+    }
+
+    pub fn send_message(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        to: NetId,
+        _svc: String,
+        sn: i64,
+        msg: Vec<u8>,
+    ) -> Result<Response, ContractError> {
+        self.ensure_xcall(deps.storage, info.sender)?;
+
+        let next_conn_sn = self.get_next_conn_sn(deps.storage)?;
+        self.store_conn_sn(deps.storage, next_conn_sn)?;
+
+        let mut fee = 0;
+
+        if sn >= 0 {
+            fee = self.get_fee(deps.storage, to.clone(), sn > 0)?.into();
+        }
+
+        let value = self.get_amount_for_denom(&info.funds, self.denom(deps.storage));
+
+        if fee > value {
+            return Err(ContractError::InsufficientFunds);
+        }
+        
+        Ok(Response::new()
+            .add_attribute("action", "send_message")
+            .add_event(
+                Event::new("Message")
+                    .add_attribute("targetNetwork", to.to_string())
+                    .add_attribute("connSn", to.to_string())
+                    .add_attribute("msg", self.hex_encode(msg)),
+            ))
+    }
+
+    pub fn recv_message(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        src_network: NetId,
+        conn_sn: u128,
+        msg: String,
+    ) -> Result<Response, ContractError> {
+        self.ensure_admin(deps.storage, info.sender)?;
+
+        let hex_string_trimmed = msg.trim_start_matches("0x");
+        let bytes = hex::decode(hex_string_trimmed).expect("Failed to decode to vec<u8>");
+        
+        let vec_msg: Vec<u8> = Binary(bytes).into();
+        let receipt = self.get_receipt(deps.as_ref().storage, src_network.clone(), conn_sn);
+        if receipt {
+            return Err(ContractError::DuplicateMessage);
+        }
+        self.store_receipt(deps.storage, src_network.clone(), conn_sn)?;
```

Comment:
`if self.get_receipt(deps.as_ref().storage, src_network.clone(), conn_sn).is_ok(){
return Err(ContractError::DuplicateMessage);
}` should be more concise.